### PR TITLE
[gree] Add support for ASC/GCM encryption

### LINKED ISSUE #16949: [gree] Unable to connect to AC
## Expected Behavior
Find the AC in the local network and control it

## Current Behavior
When starting the scan process of the Gree binding following error appear in the log:
```
2024-06-28 14:57:06.172 [DEBUG] [.internal.discovery.GreeDeviceFinder] - Sending scan packet to xx.xx.xx.255
2024-06-28 14:57:06.356 [INFO ] [ernal.discovery.GreeDiscoveryService] - Discovery: Device Discovery failed: Decryption of recieved data failed
```
When creating thing manually the error is:
```
2024-06-28 15:09:39.441 [DEBUG] [ng.gree.internal.handler.GreeHandler] - xxx: Automatic refresh started (60 second interval)
2024-06-28 15:09:39.443 [DEBUG] [.internal.discovery.GreeDeviceFinder] - Sending scan packet to xx.xx.xx.xx
2024-06-28 15:09:39.459 [INFO ] [ng.gree.internal.handler.GreeHandler] - xxx: Thing initialization failed: Decryption of recieved data failed
2024-06-28 15:09:44.442 [DEBUG] [ng.gree.internal.handler.GreeHandler] - xxx: Re-initialize device
```

## Context
New Gree devices use different encryption - ASC/GCM encyption

## Your Environment
OpenHab 4.1.1

### PR DESCRIPTION
**Pull request by lovery**
New Gree AC uses ASC/GCM encryption and different general key Fixes #16949

Fixes #16949

# Testing
Link to a built [JAR](https://file.io/EzSrZOZaFIVy) file

### CODE REVIEW

**Comment by jlaur** [comment]
@lovery - many thanks for your contribution!

@markus7017 - I know it's a stretch, but do you think you will find time for a review of this PR before the feature freeze tomorrow? 🙂

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/discovery/GreeDeviceFinder.java [hunk 1]
```diff
@@ -102,9 +102,18 @@ public void scan(DatagramSocket clientSocket, String broadcastAddress, boolean s
                         continue;
                     }
 
+                    String decryptedMsg;
                     // Decrypt message - a GreeException is thrown when something went wrong
-                    String decryptedMsg = scanResponseGson.decryptedPack = GreeCryptoUtil
-                            .decryptPack(GreeCryptoUtil.getAESGeneralKeyByteArray(), scanResponseGson.pack);
+                    // If there is a tag property in the responce then use GCM decryption
```
Comment (markus7017) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/discovery/GreeDeviceFinder.java:107]: could you move the logic to GreeCryptoUtil class, this layer here should not deal with it, just call a encrypt/decrypt method doing the details

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java [hunk 2]
```diff
@@ -117,9 +128,18 @@ public void getDeviceStatus(DatagramSocket clientSocket) throws GreeException {
             String reqStatusPackStr = GSON.toJson(reqStatusPackGson);
 
             // Encrypt and send the Status Request pack
-            String encryptedStatusReqPacket = GreeCryptoUtil.encryptPack(getKey(), reqStatusPackStr);
+            String encryptedStatusReqPacket;
+            String encryptedStatusReqTag = null;
+            if (encryptionType == EncryptionTypes.GCM) {
```
Comment (markus7017) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java:133]: could you move the logic to GreeCryptoUtil class, this layer here should not deal with it, just call a encrypt/decrypt method doing the details

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java [hunk 3]
```diff
@@ -131,7 +151,12 @@ public void getDeviceStatus(DatagramSocket clientSocket) throws GreeException {
 
             // Read the response, create the JSON to hold the response values
             GreeStatusResponseDTO resp = receiveResponse(clientSocket, GreeStatusResponseDTO.class);
-            resp.decryptedPack = GreeCryptoUtil.decryptPack(getKey(), resp.pack);
+            if (encryptionType == EncryptionTypes.GCM) {
```
Comment (markus7017) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java:154]: move logic to GreeCryptoUtil

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java [hunk 4]
```diff
@@ -157,14 +182,28 @@ public void bindWithDevice(DatagramSocket clientSocket) throws GreeException {
             String bindReqPackStr = GSON.toJson(bindReqPackGson);
 
             // Encrypt and send the Binding Request pack
-            String encryptedBindReqPacket = GreeCryptoUtil.encryptPack(GreeCryptoUtil.getAESGeneralKeyByteArray(),
-                    bindReqPackStr);
-            DatagramPacket sendPacket = createPackRequest(1, encryptedBindReqPacket);
+            String encryptedBindReqPacket;
+            String encryptedBindReqTag = null;
+            if (encryptionType == EncryptionTypes.GCM) {
```
Comment (markus7017) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java:187]: same here

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java [hunk 5]
```diff
@@ -157,14 +182,28 @@ public void bindWithDevice(DatagramSocket clientSocket) throws GreeException {
...
+                encryptedBindReqPacket = GreeCryptoUtil.encryptPack(GreeCryptoUtil.getAESGeneralKeyByteArray(),
+                        bindReqPackStr);
+            }
+            DatagramPacket sendPacket = createPackRequest(1, encryptedBindReqPacket, encryptedBindReqTag);
             clientSocket.send(sendPacket);
 
             // Recieve a response, create the JSON to hold the response values
             GreeBindResponseDTO resp = receiveResponse(clientSocket, GreeBindResponseDTO.class);
-            resp.decryptedPack = GreeCryptoUtil.decryptPack(GreeCryptoUtil.getAESGeneralKeyByteArray(), resp.pack);
+            if (encryptionType == EncryptionTypes.GCM) {
```
Comment (markus7017) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java:201]: and here

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java [hunk 6]
```diff
@@ -424,13 +463,26 @@ protected void executeCommand(DatagramSocket clientSocket, Map<String, Integer>
...
+            } else {
+                encryptedCommandReqPacket = GreeCryptoUtil.encryptPack(getKey(), execCmdPackStr);
+            }
+            DatagramPacket sendPacket = createPackRequest(0, encryptedCommandReqPacket, encryptedCommandReqTag);
             clientSocket.send(sendPacket);
 
             // Receive and decode result
             GreeExecResponseDTO execResponseGson = receiveResponse(clientSocket, GreeExecResponseDTO.class);
-            execResponseGson.decryptedPack = GreeCryptoUtil.decryptPack(getKey(), execResponseGson.pack);
+            if (encryptionType == EncryptionTypes.GCM) {
```
Comment (markus7017) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java:480]: move logic to decryptPack

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java [hunk 7]
```diff
@@ -424,13 +463,26 @@ protected void executeCommand(DatagramSocket clientSocket, Map<String, Integer>
             String execCmdPackStr = GSON.toJson(execCmdPackGson);
 
             // Now encrypt and send the Command Request pack
-            String encryptedCommandReqPacket = GreeCryptoUtil.encryptPack(getKey(), execCmdPackStr);
-            DatagramPacket sendPacket = createPackRequest(0, encryptedCommandReqPacket);
+            String encryptedCommandReqPacket;
+            String encryptedCommandReqTag = null;
+            if (encryptionType == EncryptionTypes.GCM) {
```
Comment (markus7017) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java:468]: move logic to encryptPack()

**Review by markus7017 — COMMENTED** [review]
Not essential, but you might encapsulate of ECB/GSM handling to GreeCryptoUtil, this hides details from the protocol layer

**Comment by lsiepel** [comment]
@markus7017 ping me when your finished

**Comment by lovery** [comment]
@lsiepel @markus7017, sorry for the noise, I found a few things, I didn't like and fix them today, so I think now I addressed everything in the @markus7017 review.
Unfortunately I can't test with old equipment because I don't have it, but I will ask now in the community form if someone is willing to test my changes.

**Comment by lsiepel** [comment]
> @lsiepel @markus7017, sorry for the noise, I found a few things, I didn't like and fix them today, so I think now I addressed everything in the @markus7017 review. Unfortunately I can't test with old equipment because I don't have it, but I will ask now in the community form if someone is willing to test my changes.

Great! Ping me when you think it is tested enough. If you get zero response we might need to get this into 4.3m1 so it can have some test mileages before 4.3 stable.

Sidenote: preferably don’t use git force push as we can’t track changes.

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeRequestDTO.java [hunk 8]
```diff
@@ -15,16 +15,9 @@
 /**
  *
  * The GreeBindRequest4Gson class is used by Gson to hold values to be send to
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeRequestDTO.java:17]:
```suggestion
 * The GreeRequest class is used by Gson to hold values to be send to
```

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeReqStatusPackDTO.java [hunk 9]
```diff
@@ -21,7 +21,6 @@
  * @author John Cunha - Initial contribution
  */
 public class GreeReqStatusPackDTO {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeReqStatusPackDTO.java:23]: While at it, can you adjust the javadoc to match the classname

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeReqStatusDTO.java [hunk 10]
```diff
@@ -20,11 +20,5 @@
...
  * @author John Cunha - Initial contribution
  */
-public class GreeReqStatusDTO {
-    public String cid = null;
-    public int i = 0;
-    public String t = null;
-    public int uid = 0;
-    public String pack = null;
-    public String tcid = null;
+public class GreeReqStatusDTO extends GreeRequestDTO {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeReqStatusDTO.java:23]: While at it, can you adjust the javadoc to match the class name

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeExecResponseDTO.java [hunk 11]
```diff
@@ -20,15 +20,6 @@
...
-
-    public String t = null;
-    public int i = 0;
-    public int uid = 0;
-    public String cid = null;
-    public String tcid = null;
-    public String pack = null;
-
-    public transient String decryptedPack = null;
+public class GreeExecResponseDTO extends GreeResponseBaseDTO {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeExecResponseDTO.java:23]: While at it, can you adjust the javadoc to match the class name

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeBindResponseDTO.java [hunk 12]
```diff
@@ -19,15 +19,6 @@
...
-
-    public String t = null;
-    public int i = 0;
-    public int uid = 0;
-    public String cid = null;
-    public String tcid = null;
-    public String pack = null;
-
-    public transient String decryptedPack = null;
+public class GreeBindResponseDTO extends GreeResponseBaseDTO {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeBindResponseDTO.java:22]: While at it, can you adjust the javadoc to match the class name

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeBaseDTO.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.gree.internal.gson;
+
+/**
+ *
+ * The GreeBase4Gson class is used as a base class for request and response classes
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeBaseDTO.java:17]:
```suggestion
 * The GreeBaseDTO class is used as a base class for request and response classes
```

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeScanResponseDTO.java [hunk 14]
```diff
@@ -19,13 +19,6 @@
...
-public class GreeScanResponseDTO {
-    public String t = null;
-    public int i = 0;
-    public int uid = 0;
-    public String cid = null;
-    public String tcid = null;
-    public String pack = null;
-    public transient String decryptedPack = null;
-    public transient GreeScanReponsePackDTO packJson = null;
+public class GreeScanResponseDTO extends GreeResponseBaseDTO {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeScanResponseDTO.java:22]: Please correct javadoc

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeScanResponsePackDTO.java [hunk 15]
```diff
@@ -19,8 +19,7 @@
  *
  * @author John Cunha - Initial contribution
  */
-public class GreeScanReponsePackDTO {
-
+public class GreeScanResponsePackDTO {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeScanResponsePackDTO.java:22]: Please correct javadoc method name

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeStatusResponseDTO.java [hunk 16]
```diff
@@ -20,15 +20,6 @@
...
-
-    public String t = null;
-    public int i = 0;
-    public int uid = 0;
-    public String cid = null;
-    public String tcid = null;
-    public String pack = null;
-
-    public transient String decryptedPack = null;
+public class GreeStatusResponseDTO extends GreeResponseBaseDTO {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/gson/GreeStatusResponseDTO.java:23]: Please correct javadoc method name

File: bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java [hunk 17]
```diff
@@ -451,14 +452,21 @@ private void setCommandValue(DatagramSocket clientSocket, String command, int va
...
         request.t = GREE_CMDT_PACK;
         request.uid = 0;
         request.tcid = getId();
-        request.pack = pack;
+        request.pack = data[0];
+        if (encType == GreeCryptoUtil.EncryptionTypes.GCM) {
+            if (data.length > 1) {
+                request.tag = data[1];
+            } else {
+                logger.error("Missing string for tag property for GCM encryption data");
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeAirDevice.java:467]:
```suggestion
                logger.warn("Missing string for tag property for GCM encryption data");
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Without deeper knowledge of the encryption used, i have added some minor comments.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Pull request merged**